File: app/Sources/LocalCaptionKit/Sentences.swift
import Foundation

/// Sentence utilities for the clipboard "Copy last N sentences" feature (SPEC.md §9.4).
public enum Sentences {
    /// Split text into sentences on terminal punctuation (`.`, `!`, `?`), keeping the mark.
    public static func split(_ text: String) -> [String] {
        var out: [String] = []
        var cur = ""
        for ch in text {
            cur.append(ch)
            if ch == "." || ch == "!" || ch == "?" {
                let trimmed = cur.trimmingCharacters(in: .whitespacesAndNewlines)
                if !trimmed.isEmpty { out.append(trimmed) }
                cur = ""
            }
        }
        let tail = cur.trimmingCharacters(in: .whitespacesAndNewlines)
        if !tail.isEmpty { out.append(tail) }   // trailing fragment without punctuation
        return out
    }

    /// The last `n` sentences of `text`, joined with spaces.
    public static func lastN(_ text: String, n: Int) -> String {
        guard n > 0 else { return "" }
        return split(text).suffix(n).joined(separator: " ")
    }

    /// Append a provisional caption to committed text before selecting the last `n`
    /// sentences. This lets clipboard callers expose an utterance at its endpoint without
    /// waiting for the final transcription pass.
    public static func lastN(_ text: String, appending provisional: String, n: Int) -> String {
        let committed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let temporary = provisional.trimmingCharacters(in: .whitespacesAndNewlines)
        let combined: String
        if committed.isEmpty { combined = temporary }
        else if temporary.isEmpty { combined = committed }
        else { combined = committed + " " + temporary }
        return lastN(combined, n: n)
    }
}

File: windows/src/LocalCaption.Core/Captions/Sentences.cs
using System.Text;

namespace LocalCaption.Core.Captions;

/// <summary>Sentence utilities behind the clipboard's "Copy last N sentences" (SPEC.md §9.4).</summary>
public static class Sentences
{
    /// <summary>
    /// Split on terminal punctuation (<c>.</c>, <c>!</c>, <c>?</c>), keeping the mark. A
    /// trailing fragment without punctuation is still a sentence — the live line usually is.
    /// </summary>
    public static List<string> Split(string text)
    {
        var result = new List<string>();
        var current = new StringBuilder();
        foreach (var ch in text)
        {
            current.Append(ch);
            if (ch is not ('.' or '!' or '?')) continue;
            var trimmed = current.ToString().Trim();
            if (trimmed.Length > 0) result.Add(trimmed);
            current.Clear();
        }
        var tail = current.ToString().Trim();
        if (tail.Length > 0) result.Add(tail);
        return result;
    }

    /// <summary>The last <paramref name="n"/> sentences, joined with spaces.</summary>
    public static string LastN(string text, int n)
    {
        if (n <= 0) return "";
        var all = Split(text);
        return string.Join(" ", all.Skip(Math.Max(0, all.Count - n)));
    }

    /// <summary>
    /// Append a provisional caption to the committed text before selecting. This is what
    /// lets the clipboard expose an utterance at its speech endpoint, without waiting for
    /// the final transcription pass.
    /// </summary>
    public static string LastN(string text, string provisional, int n)
    {
        var committed = text.Trim();
        var temporary = provisional.Trim();
        var combined = committed.Length == 0 ? temporary
                     : temporary.Length == 0 ? committed
                     : committed + " " + temporary;
        return LastN(combined, n);
    }
}
